File: Week 01/id_477/189.旋转数组.go
package leetcode

/*
 * @lc app=leetcode.cn id=189 lang=golang
 *
 * [189] 旋转数组
 */

// @lc code=start
func rotate(nums []int, k int) {
	//方法一：使用额外k个空间
	// if k = k % len(nums); k > 0 {
	// 	tmpArr := make([]int, 0, k)
	// 	for i := 0; i < len(nums); i++ {
	// 		if i < k {
	// 			tmpArr = append(tmpArr, nums[len(nums)-i-1])
	// 		} else {
	// 			nums[len(nums)-1-i+k] = nums[len(nums)-1-i]
	// 		}
	// 	}
	// 	for i := k - 1; i >= 0; i-- {
	// 		nums[k-i-1] = tmpArr[i]
	// 	}
	// }

	// 方法二：使用len(nums)个额外空间，简洁明了，推荐
	// lenght := len(nums)
	// new := make([]int, lenght)
	// for i := 0; i < lenght; i++ {
	// 	new[(i+k)%lenght] = nums[i]
	// }
	// copy(nums, new)

	// 方法三：反转法
	k = k % len(nums)
	reverseSlice(nums, 0, len(nums)-1)
	reverseSlice(nums, 0, k-1)
	reverseSlice(nums, k, len(nums)-1)
}

func reverseSlice(nums []int, i, j int) {
	for i < j {
		nums[i], nums[j] = nums[j], nums[i]
		i++
		j--
	}
}

// @lc code=end

File: Week 01/id_437/189.旋转数组.js
/*
 * @lc app=leetcode.cn id=189 lang=javascript
 *
 * [189] 旋转数组
 * 解提：
 * 1. 暴力破解
 * 2. 求余数
 */

// @lc code=start
/**
 * 方法一：暴力破解
 * @param {number[]} nums
 * @param {number} k
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var rotate = function(nums, k) {
    for (let i = 0; i < k; i++) {
      let previous = nums[nums.length - 1]
      for (let j = 0; j < nums.length; j++) {
        let temp = nums[j]
        nums[j] = previous
        previous = temp
      }
    }
};
// @lc code=end

// @lc code=start
/**
 * 方法二：求余数,需要使用额外的数组
 * 
 *  * @param {number[]} nums
 * @param {number} k
 * @return {void} Do not return anything, modify nums in-place instead.
 */

var rotate = function(nums, k) {
  let temp = []
  for (let i = 0; i < nums.length; i++) {
    temp[(i + k) % nums.length] = nums[i]
  }
  for (let i = 0; i < nums.length; i++) {
      nums[i] = temp[i]
  }
  return nums
}
// @lc code=end

// @lc code=start
/**
 * 方法三：环状替换
 * 
 *  * @param {number[]} nums
 * @param {number} k
 * @return {void} Do not return anything, modify nums in-place instead.
 */

var rotate = function(nums, k) {
  k = k % nums.length // k超出数组长度
  let count = 0
  for (let i = 0; count < nums.length; i++) {
    let current = i
    let prev = nums[i]
    do {
      let next = (current + k) % nums.length
      let temp = nums[next]
      nums[next] = prev
      prev = temp
      current = next
      count++
    } while (current !== i)
  }
  return nums
}
// @lc code=end
